File: jetpack_app/src/main/java/com/sty/ne/jetpack/phonecall/jetpack_app/MainActivity.java
package com.sty.ne.jetpack.phonecall.jetpack_app;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.ViewModelProvider;

import android.Manifest;
import android.os.Bundle;

import com.sty.ne.jetpack.phonecall.jetpack_app.databinding.ActivityMainBinding;
import com.sty.ne.jetpack.phonecall.jetpack_app.utils.PermissionUtils;

/**
 * Activity 很轻松，体现了单一职责原则，只做一件事情，只做管理
 */
public class MainActivity extends AppCompatActivity {
    private ActivityMainBinding binding;
    private MainViewModel viewModel;
    private String[] needPermissions = { Manifest.permission.CALL_PHONE};

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        //setContentView(R.layout.activity_main);

        //绑定管理操作
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main);

        //没有任何效果的实例化
        //viewModel = new MainViewModel(getApplication());

        //实例化(只适合 extends ViewModel)
        //viewModel = new ViewModelProvider(this, new ViewModelProvider.NewInstanceFactory()).get(MainViewModel.class);

        //实例化（适合 extends AndroidViewModel）
        viewModel = new ViewModelProvider(this, new ViewModelProvider.AndroidViewModelFactory(getApplication())).get(MainViewModel.class);
        //绑定ViewModel
        binding.setVm(viewModel);
        //建立感应绑定，如果没有的话就无法感应
        binding.setLifecycleOwner(this);

        requestPermissions();
    }

    private void requestPermissions() {
        if (!PermissionUtils.checkPermissions(this, needPermissions)) {
            PermissionUtils.requestPermissions(this, needPermissions);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        if (requestCode == PermissionUtils.REQUEST_PERMISSIONS_CODE) {
            if (!PermissionUtils.verifyPermissions(grantResults)) {
                PermissionUtils.showMissingPermissionDialog(this);
            } else {

            }
        }
    }
}

File: jetpack_app_kotlin/src/main/java/com/sty/ne/jetpack/jetpack_app_kotlin/MainActivity.kt
package com.sty.ne.jetpack.jetpack_app_kotlin

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.ViewModelProvider
import com.sty.ne.jetpack.jetpack_app_kotlin.databinding.ActivityMainBinding

/**
 * : 继承 实现
 * () 构造的概念
 * fun function函数
 * ？ 全体注意，有可能为null
 * Unit void
 */
class MainActivity : BaseActivity() {
    lateinit var binding : ActivityMainBinding
    lateinit var vm: MainViewModel

//    override fun getLayoutID(): Int {
//        return R.layout.activity_main;
//    }

    override fun getLayoutID() : Int = R.layout.activity_main


    override fun initViewAction() {
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main)
        vm = ViewModelProvider(this, ViewModelProvider.AndroidViewModelFactory(application)).get(MainViewModel::class.java)

        //建立绑定 vm + binding
        binding.vm = vm

        //建立感应关系
        binding.lifecycleOwner = this
    }

    override fun initDataAction() {
        Log.e("sty", "${method(1, 2)}")
    }

    //可以这样写函数
    val method = {number1: Int, number2: Int -> number1 + number2}

    override fun initListenerAction() {

    }

    override fun testCallbackMethod(name: String) {

    }
}
